File: src/claw_mem/security/audit.py
# Copyright 2026 Peter Cheng
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

"""
Audit Logger (MVP Version)

Records all memory operations for auditing and debugging.
"""

import json
from datetime import datetime
from typing import Dict, Optional
from pathlib import Path


class AuditLogger:
    """
    Audit Logger
    
    MVP version logs to file. Log rotation and analysis will be added in future iterations.
    """
    
    def __init__(self, workspace: Path):
        """
        Initialize Audit Logger
        
        Args:
            workspace: Workspace path
        """
        self.workspace = workspace
        self.log_file = workspace / ".audit_log.jsonl"
    
    def log(self, action: str, details: Dict) -> None:
        """
        Log audit entry
        
        Args:
            action: Action type
            details: Action details
        """
        log_entry = {
            "timestamp": datetime.now().isoformat(),
            "action": action,
            "details": details
        }
        
        # Append to log file
        with open(self.log_file, "a", encoding="utf-8") as f:
            f.write(json.dumps(log_entry, ensure_ascii=False) + "\n")
    
    def get_logs(self, action_filter: Optional[str] = None, 
                 limit: int = 100) -> list:
        """
        Get audit logs
        
        Args:
            action_filter: Action type filter
            limit: Number of results
            
        Returns:
            list: Log entries
        """
        logs = []
        
        if not self.log_file.exists():
            return logs
        
        with open(self.log_file, "r", encoding="utf-8") as f:
            for line in f:
                try:
                    log_entry = json.loads(line)
                    if action_filter is None or log_entry.get("action") == action_filter:
                        logs.append(log_entry)
                        if len(logs) >= limit:
                            break
                except json.JSONDecodeError:
                    continue
        
        return logs
    
    def clear(self) -> None:
        """
        Clear audit logs
        """
        if self.log_file.exists():
            self.log_file.unlink()

File: src/security/audit.ts
// Copyright 2026 Peter Cheng
// Licensed under Apache-2.0

/**
 * Audit Logger (MVP Version)
 *
 * Records all memory operations for auditing and debugging.
 * Appends JSONL entries to a log file in the workspace.
 */

import * as fs from "fs";
import * as path from "path";

export interface AuditEntry {
  timestamp: string;
  action: string;
  details: Record<string, unknown>;
}

export class AuditLogger {
  private workspace: string;
  private logFile: string;

  /**
   * @param workspace - Workspace directory path
   */
  constructor(workspace: string) {
    this.workspace = workspace;
    this.logFile = path.join(workspace, ".audit_log.jsonl");
  }

  /**
   * Log an audit entry.
   *
   * @param action  - Action type
   * @param details - Action details
   */
  log(action: string, details: Record<string, unknown>): void {
    const entry: AuditEntry = {
      timestamp: new Date().toISOString(),
      action,
      details,
    };

    const dir = path.dirname(this.logFile);
    fs.mkdirSync(dir, { recursive: true });
    fs.appendFileSync(this.logFile, JSON.stringify(entry) + "\n", "utf-8");
  }

  /**
   * Retrieve audit logs, optionally filtered by action type.
   *
   * @param actionFilter - Optional action type to filter by
   * @param limit        - Maximum number of entries to return (default 100)
   */
  getLogs(actionFilter?: string, limit: number = 100): AuditEntry[] {
    const logs: AuditEntry[] = [];

    if (!fs.existsSync(this.logFile)) {
      return logs;
    }

    const lines = fs.readFileSync(this.logFile, "utf-8").split("\n");
    for (const line of lines) {
      if (!line.trim()) continue;
      try {
        const entry: AuditEntry = JSON.parse(line);
        if (actionFilter === undefined || entry.action === actionFilter) {
          logs.push(entry);
          if (logs.length >= limit) break;
        }
      } catch {
        // skip malformed lines
      }
    }

    return logs;
  }

  /**
   * Clear the audit log file.
   */
  clear(): void {
    if (fs.existsSync(this.logFile)) {
      fs.unlinkSync(this.logFile);
    }
  }
}
